internal/runner/runner.go

```diff
@@ -0,0 +1,145 @@
+	Env []string
+
+	// Steps are the steps to execute.
+	Steps []*CommandStep
+
+	// Stdout is the Writer used for Stdout. If nil then os.Stdout is used.
+	Stdout io.Writer
+
+	// Stderr is the Writer used for Stderr. If nil then os.Stderr is used.
+	Stderr io.Writer
+}
+
+// CommandStep is a command that was executed by the builder.
+type CommandStep struct {
+	// Command is the command that was run.
+	Command []string `json:"command"`
+
+	// Env are the environment variables passed to the command.
+	Env []string `json:"env"`
+
+	// WorkingDir is the working directory the command was executed in.
+	WorkingDir string `json:"workingDir"`
+}
+
+// Dry returns the command steps as they would be executed by the runner
+// without actually executing the commands. This allows builders to get an
+// accurate set of steps in a trusted environment as executing commands will
+// execute untrusted code.
+func (r *CommandRunner) Dry() (steps []*CommandStep, err error) {
+	ctx := context.Background()
+	for _, step := range r.Steps {
+		var runStep *CommandStep
+		runStep, err = r.runStep(ctx, step, true)
+		if err != nil {
+			return // steps, err
+		}
+		steps = append(steps, runStep)
+	}
+	return // steps, err
+}
+
+// Run executes a series of commands and returns the steps that were executed
+// successfully. Commands are run in sequence and are expected to return a zero
+// exit status.
+//
+// Global environment variables are merged with steps environment variables in
+// the returned steps. In the case of duplicates the last occurrence has precidence.
+// Environment variables are *not* inherited from the current process.
+//
+// The returned CommandSteps should be included in the buildConfig provenance.
+// These are *not* the same as the runner commands. Env vars are sanitized, pwd
+// is changed to the absolute path, and only commands that executed
+// successfully are returned.
+func (r *CommandRunner) Run(ctx context.Context) (steps []*CommandStep, err error) {
+	for _, step := range r.Steps {
+		var runStep *CommandStep
+		runStep, err = r.runStep(ctx, step, false)
+		if err != nil {
+			return // steps, err
+		}
+		steps = append(steps, runStep)
+	}
+	return // steps, err
+}
+
+// runStep runs the build step and returns the CommandStep configuration
+// actually used to run the command. If dry is true then the CommandStep is
+// returned without executing the command.
+func (r *CommandRunner) runStep(ctx context.Context, step *CommandStep, dry bool) (*CommandStep, error) {
+	if len(step.Command) == 0 {
+		return nil, errors.New("command is empty")
+	}
+
+	name := step.Command[0]
+	args := step.Command[1:]
+
+	// Copy and merge the environment.
+	env := make([]string, len(r.Env), len(r.Env)+len(step.Env)+1)
+	copy(env, r.Env)
+	env = append(env, step.Env...)
+
+	// Set the POSIX PWD env var.
+	pwd, err := filepath.Abs(step.WorkingDir)
+	if err != nil {
+		return nil, err
+	}
+	env = append(env, "PWD="+pwd)
```

Comment:
without this, it would run in `$HOME`?

Reply:
No, but I think that some programs or scripts could depend on `PWD` being set so this is just here for completeness.